### PR TITLE
Finish converting the media repo code to async / await.

### PR DESCRIPTION
**Pull request by clokep**
This converts the remaining piece of the media repo code to async / await (the storage provider classes).

Note that 3rd-party providers may still use `Deferreds` here (see #7873).

This also adds a bunch of type hints.

I didn't see any docs to update about how to implement a storage provider.

### CODE REVIEW

File: synapse/rest/media/v1/storage_provider.py [hunk 1]
```diff
@@ -130,11 +124,11 @@ def store_file(self, path, file_info):
         if not os.path.exists(dirname):
             os.makedirs(dirname)
 
-        return defer_to_thread(
+        return await defer_to_thread(
             self.hs.get_reactor(), shutil.copyfile, primary_fname, backup_fname
         )
 
-    def fetch(self, path, file_info):
+    async def fetch(self, path, file_info):
```
Comment (anoadragon453) [inline synapse/rest/media/v1/storage_provider.py:137]: I'm confused why this needs to be an async function?

File: synapse/rest/media/v1/media_storage.py [hunk 2]
```diff
@@ -178,29 +188,23 @@ async def ensure_media_is_in_local_cache(self, file_info: FileInfo) -> str:
             # Fetch is supposed to return an Awaitable, but guard against
             # improper implementations.
             if inspect.isawaitable(res):
-                res = await res
-            if res:
-                with res:
+                res = await res  # type: ignore
+            if res:  # type: ignore
+                with res:  # type: ignore
```
Comment (anoadragon453) [inline synapse/rest/media/v1/media_storage.py:193]: This is unfortunate, but I don't know about solution here, other than perhaps marking `res` as `Any`?

File: synapse/rest/media/v1/storage_provider.py (hunk 1 above)
Comment (clokep) [reply synapse/rest/media/v1/storage_provider.py:137]: It does not *have* to be, but the API says it is async, so it is good to be consistent.

Note that the S3 provider is truly async: https://github.com/matrix-org/synapse-s3-storage-provider/blob/b4b6525be8d5af6356c11687cde09144d2ebf8a4/s3_storage_provider.py#L100-L108

File: synapse/rest/media/v1/media_storage.py (hunk 2 above)
Comment (clokep) [reply synapse/rest/media/v1/media_storage.py:193]: I had played it a bit to see if I could get it to understand what this is, but I'll try again. 👍

Comment (clokep) [reply synapse/rest/media/v1/media_storage.py]: I did a bit more experimenting and using `Any` worked! 👍 I updated the comments around this (and another location) to specify what the return type *should* be though! Thanks for the idea!

**Comment by clokep** [comment]
@anoadragon453 Thanks for the review! I think I've hit your comments, let me know how it looks now!

File: changelog.d/7947.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Convert more media code to async/await.
```
Comment (clokep) [inline changelog.d/7947.misc:1]:
```suggestion
Convert various parts of the codebase to async/await.
```

**Pull request merged**